text/0000-fn_name_macro.md

```diff
@@ -0,0 +1,61 @@
+- Feature Name: fn_name_macro
+- Start Date: 2019-11-14
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+<!-- https://github.com/rust-lang/rfcs/issues/1743 -->
+
+# Summary
+[summary]: #summary
+
+This RFC adds an additional macro, `function!`, to the `core` crate. When invoked, the macro expands to the name of the function that contains the call site.
+
+# Motivation
+[motivation]: #motivation
+
+This is a useful extension of Rust's existing debug reporting: `file!`, `line!`, `column!` and `module_path!`.
+
+For most people the name of the function is a much more immediate way to understand the context of a message than file name and line number.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+For debug information about what's happening within the program there's several useful macros that you might use. One of them is the `function!` macro, which expands to the name of the current function. If used outside of a function it causes a compilation error.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Use of the `function!` macro expands to the compiler's internal name for the function. This will generally be the name that the user wrote into the file but in the case of a closure or similar it will be something like the function's name with a unique suffix.
+
+The exact text is considered "debug" information and not subject to Rust's stability guarantee.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Doing this adds another macro to `core`.
+* This macro in particular cannot be implemented in Rust itself, it requires special support from `rustc`, as well as from all potential alternative compilers for the Rust language.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Previous discussion was within an issue of the RFCs repo: https://github.com/rust-lang/rfcs/issues/1743
+
+In summary, this is _extremely_ useful for debug purposes, and has been highly desired for nearly two years.
```

Comment:
```suggestion
In summary, this is useful for debug purposes, and has been desired for nearly two years.
```

Reply:
Another use case I'll be using it for: testing. In many of the tests I've written (and this reflects the kinds of crates I've been writing too), temporary directories and other resources are named according to the name of the test function. Not having to keep them in sync manually would be great.

Reply:
Not sure why you wrote this here, but your use case sounds like you need guarantees wrt. the output (or else your tests might suddenly start failing on an upgrade) and this RFC explicitly disclaims such guarantees.

Reply:
Maybe I should have put it on the section header line. It's more that Github only allows a single conversation per source line.

I don't need stability guarantees, just uniqueness (it also helps to disambiguate log lines in the face of rust-lang/rust#42474).